apps/ejabberd/src/rdbms/mongoose_rdbms_sup.erl

```diff
@@ -30,85 +30,90 @@
+-spec remove_pool(Pool :: mongoose_rdbms:pool()) -> ok.
+remove_pool(Pool) ->
+    ok = supervisor:terminate_child(?MODULE, Pool),
+    ok = supervisor:delete_child(?MODULE, Pool).
 
-
--spec pool_spec(Host :: ejabberd:server(), Id :: supervisor:child_id(),
-                Name :: atom(), Size :: pos_integer()) -> supervisor:child_spec().
-pool_spec(Host, Id, Name, Size) ->
-    Opts = [{workers, Size}, {worker, {mongoose_rdbms, Host}}, {pool_sup_shutdown, infinity}],
-    {Id, {wpool, start_pool, [Name, Opts]}, transient, 2000, supervisor, dynamic}.
+-spec pool_spec(Pool :: mongoose_rdbms:pool(), Size :: pos_integer()) -> supervisor:child_spec().
+pool_spec(Pool, Size) ->
+    Opts = [{workers, Size}, {worker, {mongoose_rdbms, Pool}}, {pool_sup_shutdown, infinity}],
+    {Pool, {wpool, start_pool, [pool_proc(Pool), Opts]}, transient, 2000, supervisor, dynamic}.
 
 
 -spec get_pids(ejabberd:server() | atom()) -> [pid()].
-get_pids(Host) when is_binary(Host) ->
-    get_pids(default_pool(Host));
-get_pids(PoolName) ->
-    case whereis(PoolName) of
+get_pids(HostOrPool) ->
+    PoolProc = pool_proc(HostOrPool),
+    case whereis(PoolProc) of
         undefined -> [];
         _ ->
-            Stats = wpool:stats(PoolName),
+            Stats = wpool:stats(PoolProc),
             {workers, Workers} = lists:keyfind(workers, 1, Stats),
-            [whereis(wpool_pool:worker_name(PoolName, WorkerId)) || {WorkerId, _} <- Workers]
+            [whereis(wpool_pool:worker_name(PoolProc, WorkerId)) || {WorkerId, _} <- Workers]
     end.
 
-
--spec default_pool(Host :: ejabberd:server()) -> atom().
-default_pool(Host) ->
-    gen_mod:get_module_proc(Host, ?POOL_NAME).
-
-
--spec pool_size(ejabberd:server()) -> integer().
-pool_size(Host) ->
-    case ejabberd_config:get_local_option({odbc_pool_size, Host}) of
+-spec pool_proc(mongoose_rdbms:server()) -> atom().
+pool_proc(Host) when is_binary(Host) ->
+    pool_proc(pool(Host));
+pool_proc(Pool) when is_atom(Pool) ->
+    gen_mod:get_module_proc(atom_to_list(Pool), mongoose_rdbms_pool).
+
+-spec pool(mongoose_rdbms:server()) -> mongoose_rdbms:pool().
+pool(Host) when is_binary(Host) ->
+    ejabberd_config:get_local_option_or_default({odbc_pool, Host}, ?DEFAULT_POOL_NAME);
+pool(Pool) when is_atom(Pool) ->
+    Pool.
+
+-spec pool_size(mongoose_rdbms:pool()) -> integer().
+pool_size(Pool) ->
+    case get_option(Pool, odbc_pool_size) of
         undefined ->
             ?DEFAULT_POOL_SIZE;
         Size when is_integer(Size) ->
             Size;
         InvalidSize ->
             Size = ?DEFAULT_POOL_SIZE,
             ?ERROR_MSG("Wrong odbc_pool_size definition '~p' "
-                       "for host ~p, default to ~p~n",
-                       [InvalidSize, Host, Size]),
+                       "for pool ~p, default to ~p~n",
+                       [InvalidSize, Pool, Size]),
             Size
     end.
+
+-spec get_option(mongoose_rdbms:pool(), atom()) -> term().
+get_option(Pool, Option) ->
```

Comment:
The motivation here was to avoid cyclic dependencies between modules - I would be less reluctant to allow `mongoose_rdbms_sup` call `mongoose_rdbms` if the worker logic was extracted from the latter - which I would recommend.